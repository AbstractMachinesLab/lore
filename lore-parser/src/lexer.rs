use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
pub enum Token {
    #[regex("#.*(\r\n|\n)?", |lex| lex.slice()[1..].parse())]
    Comment(String),

    #[token("as")]
    As,

    #[token("prefix")]
    Prefix,

    #[token("using")]
    Using,

    #[token("kind")]
    Kind,

    #[token("attr")]
    Attribute,

    #[token("rel")]
    Relation,

    #[token("in")]
    In,

    #[token(":")]
    Colon,

    #[regex("(\"([^\"\\\\]|\\\\.)*\")", |lex| lex.slice()[1..lex.slice().len() -1].parse())]
    LiteralString(String),

    #[regex("(\"\"\"([^\"\\\\]|\\\\.)*\"\"\")", |lex| lex.slice()[3..lex.slice().len() -3].parse())]
    MultiLineLiteralString(String),

    #[token("{")]
    OpenBrace,

    #[token("}")]
    ClosedBrace,

    #[token("/")]
    Slash,

    #[regex("([a-z0-9][a-z0-9-]*:|@|:)[a-zA-Z0-9()+,-.:=@;$_!*'%/?#]+", |lex| lex.slice().parse())]
    URI(String),

    #[regex("[a-zA-Z][a-zA-Z0-9-_]*", |lex| lex.slice().parse())]
    Text(String),

    #[regex("[0-9]+", |lex| lex.slice().parse())]
    Number(u64),

    #[error]
    #[regex(r"[ \t\n\f]+", logos::skip)]
    Error,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn comment() {
        let mut lex = Token::lexer(r#"# spotify:artist:2Hkut4rAAyrQxRdof7FVJq "#);

        assert_eq!(
            lex.next(),
            Some(Token::Comment(
                " spotify:artist:2Hkut4rAAyrQxRdof7FVJq ".to_string()
            ))
        );
    }

    #[test]
    fn literal_string_single_line() {
        let mut lex = Token::lexer(r#" "spotify:artist:2Hkut4rAAyrQxRdof7FVJq" "#);

        assert_eq!(
            lex.next(),
            Some(Token::LiteralString(
                "spotify:artist:2Hkut4rAAyrQxRdof7FVJq".to_string()
            ))
        );
    }
    #[test]
    fn literal_string_multi_line() {
        let mut lex = Token::lexer(
            r#" """
            spotify:artist:2Hkut4rAAyrQxRdof7FVJq
            more lines
            """ "#,
        );

        assert_eq!(
            lex.next(),
            Some(Token::MultiLineLiteralString("\n            spotify:artist:2Hkut4rAAyrQxRdof7FVJq\n            more lines\n            ".to_string())));
    }

    #[test]
    fn uri() {
        let mut lex = Token::lexer(r#" spotify:artist:2Hkut4rAAyrQxRdof7FVJq "#);

        assert_eq!(
            lex.next(),
            Some(Token::URI(
                "spotify:artist:2Hkut4rAAyrQxRdof7FVJq".to_string()
            ))
        );
    }

    #[test]
    fn using_namespace() {
        let mut lex = Token::lexer(r#" using spotify:schema/2021 "#);

        assert_eq!(lex.next(), Some(Token::Using));
        assert_eq!(
            lex.next(),
            Some(Token::URI("spotify:schema/2021".to_string()))
        );
    }

    #[test]
    fn prefix_uri() {
        let mut lex = Token::lexer(r#" prefix spotify:schema/2021 as @spotify "#);

        assert_eq!(lex.next(), Some(Token::Prefix));
        assert_eq!(
            lex.next(),
            Some(Token::URI("spotify:schema/2021".to_string()))
        );
        assert_eq!(lex.next(), Some(Token::As));
        assert_eq!(lex.next(), Some(Token::URI("@spotify".to_string())));
    }

    #[test]
    fn prefixed_uri() {
        let mut lex = Token::lexer(r#" @spotify/schema/2021 "#);

        assert_eq!(
            lex.next(),
            Some(Token::URI("@spotify/schema/2021".to_string()))
        );
    }

    #[test]
    fn lowercase_kind_name() {
        let mut lex = Token::lexer(r#" kind u "#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("u".to_string())));
    }

    #[test]
    fn lowercase_attr_name() {
        let mut lex = Token::lexer(r#" attr u "#);

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("u".to_string())));
    }

    #[test]
    fn empty_kind() {
        let mut lex = Token::lexer(r#" kind User "#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
    }

    #[test]
    fn empty_string() {
        let mut lex = Token::lexer(r#" "" "#);
        assert_eq!(lex.next(), Some(Token::LiteralString("".to_string())));
    }

    #[test]
    fn non_empty_string() {
        let mut lex = Token::lexer(r#" "what" "#);
        assert_eq!(lex.next(), Some(Token::LiteralString("what".to_string())));
    }

    #[test]
    fn kind_with_body() {
        let mut lex = Token::lexer(r#"kind User { }"#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }

    #[test]
    fn empty_attribute() {
        let mut lex = Token::lexer(r#"attr Name in User"#);

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::In));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
    }

    #[test]
    fn field() {
        let mut lex = Token::lexer(r#"range: integer"#);

        assert_eq!(lex.next(), Some(Token::Text("range".to_string())));
        assert_eq!(lex.next(), Some(Token::Colon));
        assert_eq!(lex.next(), Some(Token::Text("integer".to_string())));
    }

    #[test]
    fn field_block() {
        let mut lex = Token::lexer(r#" { hello world } "#);

        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::Text("hello".to_string())));
        assert_eq!(lex.next(), Some(Token::Text("world".to_string())));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }

    #[test]
    fn attr_with_fields() {
        let mut lex = Token::lexer(
            r#"
                attr Name {
                    @label/en "Name"
                    @label/es "Nombre"
                    @comment/en ""
                    @see-also    @other/entity

                    @symmetry       :symmetric
                    @reflexivity    :reflexive
                    @disjoint-with  :reflexive

                    @domain      User
                    @range       @lore/string
                    @cardinality 1
                }
        "#,
        );

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::URI("@label/en".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@label/es".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("Nombre".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@comment/en".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@see-also".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@other/entity".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@symmetry".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":symmetric".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@reflexivity".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":reflexive".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@disjoint-with".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":reflexive".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@domain".to_string())));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@range".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@lore/string".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@cardinality".to_string())));
        assert_eq!(lex.next(), Some(Token::Number(1)));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }
}
