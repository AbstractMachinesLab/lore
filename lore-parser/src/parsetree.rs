#[derive(Clone, Debug, Default, PartialEq)]
pub struct URI(pub String);

#[derive(Clone, Debug, PartialEq)]
pub enum Name {
    URI(URI),
    Alias(String),
}

#[derive(Clone, Debug, PartialEq)]
pub enum Literal {
    Number(u64),
    String(String),
    URI(URI),
}

#[derive(Clone, Debug, PartialEq)]
pub struct Field {
    name: URI,
    value: Literal,
}

#[derive(Clone, Debug, PartialEq)]
pub enum StructureItem {
    Comment(String),

    Alias { uri: URI, prefix: String },

    Directive { name: Name, value: Option<Literal> },

    Kind { name: Name },

    Attribute { name: Name, fields: Vec<Field> },
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Structure(pub Vec<StructureItem>);

impl Structure {
    pub fn items(&self) -> &Vec<StructureItem> {
        &self.0
    }
}
