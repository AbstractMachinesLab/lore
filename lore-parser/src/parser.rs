use crate::lexer::Token;
use crate::parsetree::*;
use logos::{Lexer, Logos};
use thiserror::Error;

#[derive(Error, Debug, PartialEq, Eq)]
pub enum ParseError {
    #[error("Expected a URI.")]
    ExpectedURI,

    #[error("The `use` syntax expects a URI.")]
    UseExpectsURI,

    #[error(
        "The `use` syntax should follow the format: use <uri> as <name>. Did you forget the 'as'?"
    )]
    UseIsMissingTheAsKeyword,

    #[error(r#"
        The `use` syntax should follow the format: use <uri> as <name>. Did you forget to specify a name?
    "#)]
    UseIsMissingTheAliasedName,

    #[error("")]
    UseExpectsTextPrefix,

    #[error("The `kind <name>` syntax is missing a name.")]
    KindIsMissingAName,

    #[error("We expected to find a Name (an Alias or a URI).")]
    NameIsInvalid,

    #[error("Runtime error")]
    Runtime(String),
}

pub fn parse(text: &str) -> Result<Structure, ParseError> {
    let mut lex = Token::lexer(text);
    parse_structure(&mut lex)
}

fn parse_structure(mut lex: &mut Lexer<Token>) -> Result<Structure, ParseError> {
    let mut items = vec![];
    while let Some(token) = lex.next() {
        let item = match token {
            Token::Use => parse_alias(&mut lex)?,
            Token::Kind => parse_kind(&mut lex)?,
            Token::Attribute => parse_attr(&mut lex)?,
            _ => continue,
        };
        items.push(item);
    }
    Ok(Structure(items))
}

fn parse_uri(lex: &mut Lexer<Token>) -> Result<URI, ParseError> {
    match lex.next() {
        Some(Token::URI(uri)) => Ok(URI(uri)),
        _ => Err(ParseError::ExpectedURI),
    }
}

fn parse_name(lex: &mut Lexer<Token>) -> Result<Name, ParseError> {
    match lex.next() {
        Some(Token::URI(uri)) => Ok(Name::URI(URI(uri))),
        Some(Token::Text(text)) => Ok(Name::Alias(text)),
        _ => Err(ParseError::NameIsInvalid),
    }
}

fn parse_alias(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let uri = match parse_uri(&mut lex) {
        Ok(uri) => Ok(uri),
        _ => Err(ParseError::UseExpectsURI),
    }?;

    match lex.next() {
        Some(Token::As) => Ok(()),
        _ => Err(ParseError::UseIsMissingTheAsKeyword),
    }?;

    let prefix = match lex.next() {
        Some(Token::Text(prefix)) => Ok(prefix),
        _ => Err(ParseError::UseIsMissingTheAliasedName),
    }?;

    Ok(StructureItem::Alias { uri, prefix })
}

fn parse_kind(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let name = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::KindIsMissingAName),
    }?;

    Ok(StructureItem::Kind { name })
}

fn parse_attr(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let name = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::KindIsMissingAName),
    }?;

    let fields = vec![];

    Ok(StructureItem::Attribute { name, fields })
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::*;

    macro_rules! test {
        ($name:ident, $src:expr) => {
            #[test]
            fn $name() {
                let result = parse($src);
                let snapshot = format!(
                    r#"
input:
    {}

output:

{:#?}
"#,
                    $src, result
                );
                assert_snapshot!(snapshot)
            }
        };
    }

    test!(
        parse_alias,
        r#" use spotify:artist:2Hkut4rAAyrQxRdof7FVJq as Rush "#
    );

    test!(
        parse_alias_with_something_that_isnt_a_uri,
        r#" use 2Hkut4rAAyrQxRdof7FVJq as Rush "#
    );

    test!(parse_alias_with_a_missing_uri, r#" use as Rush "#);

    test!(
        parse_alias_with_uri_but_missing_keyword_as,
        r#" use spotify:hello Rush "#
    );

    test!(
        parse_alias_with_uri_but_missing_the_aliased_name,
        r#" use spotify:hello as"#
    );

    test!(parse_kind_with_uri_name, r#"kind spotify:kind:artist"#);

    test!(parse_kind_with_aliased_name, r#"kind Hello"#);

    test!(parse_kind_with_missing_name, r#"kind"#);

    test!(parse_attr_with_uri_name, r#"attr spotify:field:Name"#);

    test!(parse_attr_without_aliased_name, r#"attr Name"#);

    test!(
        parse_multiple_items,
        r#"
            use spotify:kind:artist as Artist
            kind Artist
            attr Name
            attr spotify:field:play_count
            kind spotify:kind:Album
            use spotify:kind:song as Song
            kind Song
        "#
    );
}
