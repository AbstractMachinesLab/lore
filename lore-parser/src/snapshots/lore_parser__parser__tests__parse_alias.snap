---
source: lore-parser/src/parser.rs
expression: snapshot

---

input:
     use spotify:artist:2Hkut4rAAyrQxRdof7FVJq as Rush 

output:

Ok(
    Structure(
        [
            Alias {
                uri: URI(
                    "spotify:artist:2Hkut4rAAyrQxRdof7FVJq",
                ),
                prefix: "Rush",
            },
        ],
    ),
)

